### ISSUE TITLE
[Road to 3Box Full] Blocking Issues

### ISSUE THREAD

**Issue by kuhnchris** [3box/3box-react-native#4]
Here I'm writing my noticed blockers for full-3box-rn release:
WORK IN PROGRESS; List may change drastically over time

* IPFS (and ecosystem)
  * [high, crypto] generateKeys in crypto, webcrypto and webcrypto.subtle in JWK not working properly on RN
**Attempts**: 
    * webcrypto-shim - not working
    * msrcrypto - "algorithm not supported"
    * react-native-crypto - not working
  * [high, storage] webStorage/indexedDB not working 
**Solution**:  shimmed via indexeddbshim & react-native-sqlite
  * [low, webstandards] AbortController not working properly
**Solution(?)**: shimming via abort-controller
  * [low, encoding] btoa & atob missing
**Solution**: shimming via base64 encode/decode
  * [med, webstandards] global.navigator, global.navigator.userAgent, global.location, document.createElement
**Solution**: hardcoded values, the libs in question should be refactored...
  * [high, buffer] Buffer not found
**Solution**: rn-nodeify
  * [high, WantManager] using libp2p-tcp, we are no able to add any peer to the want-manager
**Info**: We can connet to defined nodes (it errors out when the IPFS part of the multihash is incorrect), we do get initial "data" and "connect" but they are never used as peers in the WantManager

* libP2P
  * [high, tcp] libp2p-tcp not working
**partial solution**: using react-native-tcp, multihash has a bug concerning ports as string (see: https://github.com/multiformats/js-multiaddr/issues/86)

**Comment by sirpy** [comment]
@kuhnchris 
Hey.
I've helped released a working version for a webcrypto-bridge for react-native
it works by using a webview to call chrome/safari native webcrypto
check it out https://github.com/webview-crypto/react-native-webview-crypto
What other issues are blocking 3box from running on react-native?

**Comment by kuhnchris** [comment]
Hey there @sirpy. I haven't worked on this for nearly a year now - surely the landscape will have changed now and needs to be reevaluated by the 3box team. :-)

**Comment by x5engine** [comment]
Would love to see this working :)

**Comment by oed** [comment]
The main blocker is getting this to run in RN: https://github.com/openworklabs/react-native-ipfs-http-client
We have been working on this as a side project, but it's paused for now.

**Comment by sirpy** [comment]
@oed what are the main blockers there?

**Comment by oed** [comment]
@sirpy I believe there was some blockers from the js-ipfs side. Not sure exactly what those are.

Textile is written in go and uses a go-ipfs node. They are currently reworking their model to have a model similar to js-ipfs-http-client as well. 

3Box requires dag get/put as well as ipfs pubsub in order to function properly :)

**Comment by kuhnchris** [comment]
exactly, and the DAG was the reason we currently failed. We were able to enable a basic P2P stream but all the libraries in conjunction started using crypto implementation that are currently barely implemented in the native-react environment, or only stubs, or only working with various hacks applied to each single library on it's own.
After about 1 or 2 weeks of trying around I basically gave up until we have a better alternative to the various libp2p-transport layers and the various orchistrated sub systems required. :-)

But - if you want to give it a try @sirpy - you could always give it a try (or jump onto the 3box discord to discuss futher steps. :-) ).

**Comment by sirpy** [comment]
maybe the easiest solution is to also make a 3box webview bridge
@kuhnchris @oed

**Comment by oed** [comment]
@sirpy yes that possible. Actually Metamask already does this for their app.

**Comment by tesla809** [comment]
Any progress on this yet?

**Comment by x5engine** [comment]
fund this and let me knock my head at it ;)

**Comment by oed** [comment]
This issue is mainly being tracked here: https://github.com/ipfs/js-ipfs/issues/2813
